File: scraping/models/clubs.py
'''

the club model

'''

from sqlalchemy import Float, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from models.base_model import BaseModel


class Club(BaseModel):
    '''the club model'''

    __tablename__ = 'clubs'
    name: Mapped[str] = mapped_column(String, nullable=False, unique=True)
    founded: Mapped[int] = mapped_column(Integer, nullable=False)
    players = relationship('Player', back_populates='club')
    city: Mapped[str] = mapped_column(String, nullable=True)
    position: Mapped[int] = mapped_column(Integer, nullable=False)
    matches_played: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0
    )
    wins: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    draws: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    losses: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    goal_difference: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0
    )
    goals_scored: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0
    )
    goals_conceded: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0
    )
    points: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    ppg: Mapped[float] = mapped_column(Float, nullable=False, default=0)
    xG: Mapped[float] = mapped_column(Float, nullable=False, default=0)
    xGA: Mapped[float] = mapped_column(Float, nullable=False, default=0)
    xGD: Mapped[float] = mapped_column(Float, nullable=False, default=0)
    xGDp90: Mapped[float] = mapped_column(Float, nullable=False, default=0)
    xGp90: Mapped[float] = mapped_column(Float, nullable=False, default=0)
    last_five: Mapped[str] = mapped_column(String, nullable=True, default='')
    attendance_per_game: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0
    )
    possession: Mapped[float] = mapped_column(Float, nullable=True)
    prgC: Mapped[int] = mapped_column(Integer, default=0, nullable=True)
    prgP: Mapped[int] = mapped_column(Integer, default=0, nullable=True)
    no_players: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    average_age: Mapped[float] = mapped_column(
        Float, default=0.0, nullable=False
    )
    stadium: Mapped[str] = mapped_column(String, nullable=True)
    nickname: Mapped[str] = mapped_column(String, nullable=True)
    stadium_capacity: Mapped[int] = mapped_column(Integer, nullable=True)
    manager: Mapped[str] = mapped_column(String, nullable=True)

    def to_dict(self, include: list = None, exclude: list = None):
        '''to_dict method'''
        data = {
            col.name: getattr(self, col.name) for col in self.__table__.columns
        }
        if include:
            data = {key: data[key] for key in include if key in data}
        if exclude:
            for key in exclude:
                data.pop(key, None)
        return data

File: api/src/models/clubs.js
import {
  pgTable, unique, varchar, integer, doublePrecision, serial, timestamp,
} from 'drizzle-orm/pg-core';

const clubs = pgTable('clubs', {
  name: varchar().notNull(),
  founded: integer().notNull(),
  city: varchar(),
  position: integer().notNull(),
  matchesPlayed: integer('matches_played').notNull(),
  wins: integer().notNull(),
  draws: integer().notNull(),
  losses: integer().notNull(),
  goalDifference: integer('goal_difference').notNull(),
  goalsScored: integer('goals_scored').notNull(),
  goalsConceded: integer('goals_conceded').notNull(),
  points: integer().notNull(),
  ppg: doublePrecision().notNull(),
  xG: doublePrecision().notNull(),
  xGA: doublePrecision().notNull(),
  xGD: doublePrecision().notNull(),
  xGDp90: doublePrecision().notNull(),
  xGp90: doublePrecision().notNull(),
  lastFive: varchar('last_five'),
  attendancePerGame: integer('attendance_per_game').notNull(),
  possession: doublePrecision(),
  prgC: integer(),
  prgP: integer(),
  noPlayers: integer('no_players').notNull(),
  averageAge: doublePrecision('average_age').notNull(),
  stadium: varchar(),
  nickname: varchar(),
  stadiumCapacity: integer('stadium_capacity'),
  manager: varchar(),
  id: serial().primaryKey().notNull(),
  createdAt: timestamp('created_at', { mode: 'string' }).notNull().defaultNow(),
  updatedAt: timestamp('updated_at', { mode: 'string' }).notNull(),
}, (table) => [
  unique('clubs_name_key').on(table.name),
]);

export default clubs;
